src/main/java/com/tabnine/TabNineProcess.java

```diff
@@ -2,37 +2,62 @@
 
 import com.google.gson.Gson;
 import com.google.gson.GsonBuilder;
-import com.intellij.openapi.diagnostic.Logger;
 import com.intellij.openapi.application.ApplicationInfo;
+import com.intellij.openapi.diagnostic.Logger;
 import com.intellij.util.PlatformUtils;
+import org.jetbrains.annotations.NotNull;
 
-import java.io.BufferedReader;
 import java.io.IOException;
-import java.io.InputStreamReader;
 import java.nio.charset.StandardCharsets;
-import java.util.ArrayList;
-import java.util.HashMap;
-import java.util.List;
-import java.util.Map;
+import java.util.*;
+
+import static com.tabnine.Utils.emptyUponException;
+import static java.util.Collections.singletonList;
+import static java.util.Collections.singletonMap;
 
 class TabNineProcess {
-    Process proc;
-    BufferedReader procLineReader;
-    int restartCount = 0;
+    public static final String BINARY_VERSION = "2.0.2";
 
-    TabNineProcess() throws IOException {
-        this.startTabNine();
+    public TabNineProcess() throws IOException {
+        TabNineProcessFactory.create(generateCommand());
     }
 
-    void startTabNine() throws IOException {
-        if (this.proc != null) {
-            this.proc.destroy();
-            this.proc = null;
+    public <T> T request(Request<T> request) {
+        Gson gson = new GsonBuilder().create();
+        Map<String, Object> jsonObject = new HashMap<>();
+        jsonObject.put("version", BINARY_VERSION);
```

Comment:
I realize it is not new, but why are we passing "2.0.2" as version property?

Reply:
This is the binary's protocol version

Reply:
I'll rename it